sakai-evaluation-impl/src/java/org/sakaiproject/evaluation/logic/externals/ExternalHierarchyLogicImpl.java

```diff
@@ -267,56 +497,152 @@ public void setEvalGroupsForNode(String nodeId, Set<String> evalGroupIds) {
         if (evalHierarchyProvider != null) {
             l = evalHierarchyProvider.getNodesAboveEvalGroup(evalGroupId);
-        } else {
-            String nodeId = dao.getNodeIdForEvalGroup(evalGroupId);
-            if (nodeId != null) {
-                HierarchyNode currentNode = hierarchyService.getNodeById(nodeId);
-                Set<HierarchyNode> parents = hierarchyService.getParentNodes(nodeId, false);
-                parents.add(currentNode);
-                List<HierarchyNode> sorted = HierarchyUtils.getSortedNodes(parents);
-                // now convert the nodes to eval nodes
-                for (HierarchyNode node : sorted) {
-                    l.add( makeEvalNode(node) );
+        }
+
+        // Support for resolving eval groups based on hierarchy node rules
+        // Hierarchy rules should be obeyed regardless of if an external provider is present or not (supplemental)
+        String nodeID = "";
+        String siteID = evalGroupId.replace( EvalConstants.GROUP_ID_SITE_PREFIX, "" );
+        boolean isGroupIDSectionBased = false;
+        if( evalGroupId.contains( EvalConstants.GROUP_ID_SECTION_PREFIX ) )
+        {
+            siteID = siteID.substring( 0, siteID.indexOf( EvalConstants.GROUP_ID_SECTION_PREFIX ) );
+            isGroupIDSectionBased = true;
+        }
+
+        try
+        {
+            Site site = siteService.getSite( siteID );
+            List<HierarchyNodeRule> rules = externalLogic.getAllRules();
+            for( HierarchyNodeRule rule : rules )
+            {
+                String rawRuleText = removeQualifierFromRuleText( rule.getRule() );
+                String qualifier = determineQualifierFromRuleText( rule.getRule() );
+
+                // If the rule is section based and so is the group ID...
+                if( EvalConstants.HIERARCHY_RULE_SECTION.equals( rule.getOption() ) && isGroupIDSectionBased )
+                {
+                    // Get the sections of the site
+                    String realmID = siteService.siteReference( site.getId() );
+                    Set<String> sectionIDs = authzGroupService.getProviderIds( realmID );
+                    for( String sectionID : sectionIDs )
+                    {
+                        Section section = courseManagementService.getSection( sectionID );
+
+                        // If the qualifier is 'contains' and the section title contains the raw rule text, OR
+                        // the qualifier is 'ends with' and the section title ends with the raw rule text, OR
+                        // the qualifier is 'starts with' and the section title starts with the raw rule text, OR
+                        // the qualifier is 'is' and the section title equals the raw rule text, grab the node ID of the rule
+                        if( (EvalConstants.HIERARCHY_QUALIFIER_CONTAINS.equals( qualifier ) && section.getTitle().contains( rawRuleText ))
```

Comment:
You can create a method that takes String title and put the comparison logic in there. That way you can use it again for site titles since the comparisons are identical.